### PR TITLE
KAFKA-13592:Fix flaky test ControllerIntegrationTest.testTopicIdUpgradeAfterReassigningPartitions

### PR DESCRIPTION
**Pull request by Kvicii**
this issue similar to https://github.com/apache/kafka/pull/11666.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dajac** [comment]
@Kvicii Thanks for the PR. I am a bit confused because the PR refactors the code but it does not seem to fix any issues. Did I miss something? Have you been able to make the test fail?

**Comment by Kvicii** [comment]
> 

The first time it failed due to other tests, so I rebuilt it. During the build I see that testTopicIdCreatedOnUpgrade has passed.

**Comment by Kvicii** [comment]
@dajac The first build failed didn't seem to be because of this test, so I rebuilt it.Can you tell me what the expectations should be, thanks a lot

**Comment by dajac** [comment]
@Kvicii `ControllerIntegrationTest.testTopicIdUpgradeAfterReassigningPartitions` does not fail all the time but only occasionally. It seems that there is a race condition in the test which causes it. I suggest to first try to reproduce the failure by running the test in a while loop. You can do the following:

```
while ./gradlew cleanTest :core:test --tests ControllerIntegrationTest.testTopicIdUpgradeAfterReassigningPartitions; do :; done;
```

Once we understand the issue, we can improve the test to fix it.

**Comment by Kvicii** [comment]
@dajac I used `computeUntilTrue`, but it is also very difficult to reproduce using a script.I tried a few other ways, but none seem to work😔

**Comment by divijvaidya** [comment]
@Kvicii can you please fix the conflicts on this one? I think that you have the correct fix here and once you have fixed the conflicts, we can ask for a review from the community.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: A similar change is required for other tests in this file such as in `testTopicIdMigrationAndHandling`. We will need this computeUntilTrue whenever we are bringing up a new controller. Please update the other tests as well.

Comment (Kvicii) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: only `ControllerIntegrationTest.scala` file ? I will modify this.

Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Yes please. Let's just modify tests in this file for the scope of this PR. Please add a comment when you are done and we can tag one of the committers to take a look at it.

**Comment by hachikuji** [comment]
@Kvicii Thanks for the patch. I checked this out and `ControllerIntegrationTest.testTopicIdUpgradeAfterReassigningPartitions` still fails 5-10% of the time when run locally. The new failure looks like this:
```
org.opentest4j.AssertionFailedError: expected same topic ID but it can not be found ==> 
Expected :Some(6_zllbH4TES3yUSJZ4ATsw)
Actual   :None
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1152)
	at kafka.controller.ControllerIntegrationTest.testTopicIdUpgradeAfterReassigningPartitions(ControllerIntegrationTest.scala:1473)
```
I think we need to account for a delay between when the change hits zk and when it is reflected in the `ControllerContext`. I've pushed a patch here on top of this branch: https://github.com/apache/kafka/commit/f2dd2f5558d56a7d2f667d71b4c3453466dc3df8. This logic first waits for a consistent `ControllerContext`, then asserts zookeeper state. I wasn't able to reproduce the failure anymore with this patch. Feel free to include it here.

Also, could we move the unrelated cleanups into a separate PR? I know it can be difficult to overlook messy code, but it makes review more difficult since we have to look at every line to see which changes actually matter. And, on the other hand, it is easy to review cleanup PRs if they are not changing any logic.

**Comment by Kvicii** [comment]
@hachikuji ok, let me split other PR.

**Comment by hachikuji** [comment]
@Kvicii What did you think about the changes here https://github.com/apache/kafka/commit/f2dd2f5558d56a7d2f667d71b4c3453466dc3df8? I think we need something like this to get rid of some remaining race conditions in the test.

**Comment by Kvicii** [comment]
@hachikuji I agree with you.If [f2dd2f5](https://github.com/apache/kafka/commit/f2dd2f5558d56a7d2f667d71b4c3453466dc3df8) make sure the unit test pass.

**Comment by hachikuji** [comment]
@Kvicii Thanks for checking. Do you want to pull those changes here or shall I do a separate PR?

**Comment by Kvicii** [comment]
@hachikuji I moved those changes here, thx.

**Review by hachikuji — APPROVED** [review]
Thanks, LGTM.

**Pull request merged**
